pkg/wait/wait_test.go

```diff
@@ -1,26 +1,338 @@
 package wait_test
 
 import (
+	"context"
 	"errors"
+	"fmt"
+	"sync/atomic"
 	"testing"
 	"time"
 
 	. "github.com/onsi/ginkgo"
 	. "github.com/onsi/gomega"
+	"github.com/submariner-io/armada/pkg/defaults"
 	"github.com/submariner-io/armada/pkg/wait"
-
-	"sync/atomic"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/kubernetes/scheme"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
 )
 
 func TestWait(t *testing.T) {
 	RegisterFailHandler(Fail)
 	RunSpecs(t, "Wait test suite")
 }
 
+type testClient struct {
+	client.Client
+	returnError error
+}
+
+func (c *testClient) Get(ctx context.Context, key client.ObjectKey, obj runtime.Object) error {
+	err := c.returnError
+	if err != nil {
+		c.returnError = nil
+		return err
+	}
+
+	return c.Client.Get(ctx, key, obj)
+}
+
+func (c *testClient) List(ctx context.Context, list runtime.Object, opts ...client.ListOption) error {
+	err := c.returnError
+	if err != nil {
+		c.returnError = nil
+		return err
+	}
+
+	return c.Client.List(ctx, list, opts...)
+}
+
 var _ = Describe("Wait tests", func() {
+	BeforeEach(func() {
+		defaults.WaitDurationResources = 10 * time.Second
+		defaults.WaitRetryPeriod = 200 * time.Millisecond
+	})
+
 	Context("ForTasksComplete", testForTasksComplete)
+	Context("ForDeploymentReady", testForDeploymentReady)
+	Context("ForDaemonSetReady", testForDaemonSetReady)
+	Context("ForPodsRunning", testForPodsRunning)
 })
 
+func testForPodsRunning() {
+	namespace := "test-namespace"
+
+	var (
+		initialPods   []*corev1.Pod
+		client        *testClient
+		waitCh        chan error
+		listError     error
+		numReplicas   int
+		labelSelector string
+	)
+
+	BeforeEach(func() {
+		listError = nil
+		initialPods = nil
+		numReplicas = 1
+		labelSelector = "app=test"
+	})
+
+	JustBeforeEach(func() {
+		var initObjs []runtime.Object
+		for _, p := range initialPods {
+			initObjs = append(initObjs, p)
+		}
+
+		client = &testClient{Client: fake.NewFakeClientWithScheme(scheme.Scheme, initObjs...), returnError: listError}
+
+		waitCh = runAsync(func() error {
+			return wait.ForPodsRunning("east", client, namespace, labelSelector, numReplicas)
+		})
+	})
+
+	newPod := func(name string, phase corev1.PodPhase) *corev1.Pod {
```

Comment:
Seems like you only use the namespace which is constant, so why not define it as a constant for all the test and declare this function on the outside (next to runAsync for example?)

Reply:
It's only used within this function hence I scoped it as such. It doesn't really matter to me either way - if you prefer it outside I can move it.

Reply:
Whatever you prefer.
Personally I try not to use functions like this in the tests unless necessary, but if you like this style then keep it.

Reply:
I moved it.